add (short) link functionality for event

The-Ludwig opened the pull request:
Fixes #82
- default short link is a sanitized version of event name
- db migration also introduces default short links for past events
- alternative routes are neatly introduced by a custom decorator, removing the need to define a new function for each duplicate route. *SUGAR*

The-Ludwig commented:
Found a bug when creating a new view, should be fixed now

maxnoe commented on member_database/admin_views.py:
```diff
@@ -58,13 +59,14 @@ def inaccessible_callback(self, name, **kwargs):
     access_level = 'event_admin'
     column_list = ['name', 'notify_email', 'max_participants', 'force_tu_mail',
-                   'registration_open']
+                   'registration_open', Event.shortlink]
```
Any reason for using the class attribute and not just a link? Better be consistent. If there is an advantage of using the member, let's use it for all columns.

The-Ludwig replied:
No. I thought using the member has advantages, but I don't think it has. (I personally thing this is better anyway, due to autocompletion.)

maxnoe commented on member_database/admin_views.py:
```diff
@@ -58,13 +59,14 @@ def inaccessible_callback(self, name, **kwargs):
                       'registration_open', 'notify_email']
     form_excluded_columns = ['registrations']
-    column_editable_list = ['name', 'registration_open']
+    column_editable_list = ['name', 'registration_open', 'shortlink']
```
Here it's the string again

maxnoe commented on member_database/events/__init__.py:
```diff
@@ -115,7 +116,34 @@ def get_free_places(event):
     return None
 
 
+def route_name_to_id(route, **options):
```
Nice use of decorators!

maxnoe commented on member_database/events/__init__.py:
```diff
@@ -115,7 +116,34 @@ def get_free_places(event):
+        def call_with_id(name):
+            # It'd be better to add a `.order_by(Event.date)`, but 
+            # events currently are not associated with a date (new github issue?)
+            event = Event.query.filter_by(shortlink=name).first()
```
I'd rather have the shortlink have a unique constrained, that will also make lookups fast (create an index)

The-Ludwig replied:
Ah true, I didn't think about indexing. Reasoning was: If I use unique constraint here, It is not guaranteed that the DB migration will work with automatically creating a shortlink.

maxnoe commented on member_database/events/models.py:
```diff
@@ -12,6 +13,16 @@ class Event(db.Model):
     force_tu_mail = db.Column(db.Boolean, default=False)
 
+    
+    def default_shortlink(context):
```
I don't think we should generate default short links to be honest.

maxnoe commented on migrations/versions/66b173ff9a64_add_shortlink_to_event.py:
```diff
@@ -0,0 +1,37 @@
+def upgrade():
+    # ### commands auto generated by Alembic - please adjust! ###
+    with op.batch_alter_table('event', schema=None) as batch_op:
+        batch_op.add_column(sa.Column('shortlink', sa.Text(), nullable=True, default=default_shortlink))
```
I'd also not create the default shortlink for all past events. Let's just create  them by hand. E.g. we anyways would want "sommerakademie2023`, but probably just `soak23`, same for something like "bewerbungsworkshop"

The-Ludwig replied:
Yes, it's kind of an optional feature. I realize it's called shortlink, bu it's nice to have named links anyway, isn't it? No matter if they are short or not. If we want a real short link we can modify it in the admin view after all.

maxnoe commented on member_database/events/models.py:
```diff
@@ -12,6 +13,16 @@ class Event(db.Model):
+        else:
+            return ""
+
+    shortlink = db.Column(db.Text, default=default_shortlink)
```
I'd remove the default and add a unique constraint

The-Ludwig replied:
I'll add the unique constraint.

The-Ludwig replied:
My suggestion: Don't create the default names in migration, since this is not necessary compatible to the unique constraint. 
But let the default shortlink still be created. This only adds features, not remove any. If you want a non-default shortlink, just type it in the admin view!

The-Ludwig replied:
@maxnoe what do you think?

The-Ludwig commented:
I implemented your suggestions. I am unsure if the query should change with the unique contraint, or if filter_by is still the way to go.  I accidentilly run black... Should I revert?

maxnoe commented on member_database/admin_views.py:
```diff
@@ -38,13 +38,12 @@ class AuthorizedView(ModelView):
-        return (
-            current_user.is_authenticated
-            and current_user.has_access(self.access_level)
+        return current_user.is_authenticated and current_user.has_access(
```
Bad black... better add outer parentheses so it reads:

```
return (
    current_user.is_authenticated
    and current_user.has_access(self.access_level)
)
```
```

The-Ludwig commented on member_database/admin_views.py:
```diff
@@ -37,13 +38,12 @@ class AuthorizedView(ModelView):
-        return (
-            current_user.is_authenticated
-            and current_user.has_access(self.access_level)
+        return current_user.is_authenticated and current_user.has_access(
+            self.access_level
         )
```
```suggestion
        return (
            current_user.is_authenticated 
            and current_user.has_access(self.access_level)
        )
```

The-Ludwig commented:
@maxnoe I hope this should be ready to merge. 
I added your suggestions, except that I still make a default "short"link available, which can be deactivated if not wanted by the admin (check out the information on the shortlink field in the admin view). I also changed the db type from text to string.

The-Ludwig commented:
Oh and, if a shortlink is available the event listing refers to the short link, rather than the id link (via new template function `url_for_event`)

maxnoe commented:
Can you rebase this vs. master? To make the black changes and the merge commits disappear here

maxnoe commented on member_database/events/__init__.py:
```diff
@@ -326,6 +365,7 @@ def participants(event_id):
 
 
 @events.route('/<int:event_id>/write_mail/', methods=['GET', 'POST'])
+@route_name_to_id('/<string:name>/write_mail/', methods=['GET', 'POST'])
```
More magic... but would it be possible to have the decorate not need the arguments? E.g. like this:

```
@route_name_to_id
@events..route('/<int:event_id>/write_mail/', methods=['GET', 'POST'])
```

would somehow get the route and the options and do something like `route.replace("int:event_id", "string:name")`?

The-Ludwig replied:
Wondered the same thing! 
Should be possible... flask registers the function names so there is definitely a method like get get_route_rule(function) :D

The-Ludwig replied:
@maxnoe events.url_map.iter_rules() should do the trick...

The-Ludwig commented on member_database/events/__init__.py:
```diff
@@ -1,4 +1,4 @@
-from flask import (
+rom flask import (
```
```suggestion
from flask import (
```
I wonder how this got there...
